packages/collector/Collector.php

```diff
@@ -0,0 +1,229 @@
+<?php
+/**
+ * @package Collector
+ * @version 0.0.0
+ */
+/*
+Plugin Name: Collector
+Plugin URI: https://github.com/seanmorris/collector
+Description: Packages your WordPress install and sends it to Playground.
+Author: Sean Morris
+Version: 0.0.0
+Author URI: https://github.com/seanmorris/
+*/
+
+const COLLECTOR_DOWNLOAD_PATH   = '/wp-admin/?page=collector_download_package';
+const COLLECTOR_FINAL_ZIP       = '/tmp/collector-package.zip';
+
+define('COLLECTOR_PLAYGROUND_PACKAGE', ($_SERVER['SERVER_NAME'] === 'localhost')
+	? 'http://localhost:8081/index.js'
+	: 'https://unpkg.com/@wp-playground/client/index.js'
+);
+
+global $wp_version;
+
+define('COLLECTOR_WP_VERSION', $wp_version);
+define('COLLECTOR_PHP_VERSION', implode('.',sscanf(phpversion(), '%d.%d')));
+
+require __DIR__ . '/Collector_Content.php';
+require __DIR__ . '/Collector_Db.php';
+require __DIR__ . '/Collector_Helpers.php';
+require __DIR__ . '/Collector_Zip.php';
+
+add_action('admin_menu', 'collector_plugin_menu');
+add_action('plugins_loaded', 'collector_plugins_loaded');
+add_filter('plugin_install_action_links', 'collector_plugin_install_action_links', 10, 2);
+
+function collector_plugins_loaded()
+{
+    if(!current_user_can('manage_options'))
+    {
+        return;
+    }
+
+    if(urldecode($_SERVER['REQUEST_URI']) === COLLECTOR_DOWNLOAD_PATH)
+    {
+        collector_zip_collect();
+        collector_zip_send();
+        collector_zip_delete();
+        exit();
+    }
+}
+
+function collector_plugin_menu()
+{
+    add_submenu_page(
+        NULL,
+        'Collector',
+        'Collector',
+        'manage_options',
+        'collector_render_playground_page',
+        'collector_render_playground_page',
+        NULL
+    );
+}
+
+function collector_render_playground_page()
+{?>
+	<div id = "wp-playground-wrapper">
+		<div id = "wp-playground-toolbar">
+			NOW WORKING INSIDE WORDPRESS PLAYGROUND &nbsp; [<a href = "/wp-admin" id = "goBack">RETURN</a>]
+		</div>
+		<div id = "wp-playground-main-area">
+			<iframe id = "wp-playground"></iframe>
+		</div>
+	</div>
+	<script type = "text/javascript">
+		const frame  = document.getElementById('wp-playground');
+		const zipUrl = <?=json_encode(COLLECTOR_DOWNLOAD_PATH);?>;
+
+		const username   = <?=json_encode(wp_get_current_user()->user_login);?>;
```

Comment:
https://3v4l.org/IcF7CA
```php
<?php
echo json_encode('</script><script>alert("I escaped");');
```
```
Output for 8.1.7 - 8.1.24, 8.2.0 - 8.2.11
"<\/script><script>alert(\"I escaped\");"
```

Reply:
I feel uneasy about the `\/` escaping as the only barrier here. Is there any escaping function WordPress ships that would also escape the `<` and `>` characters? Maybe `esc_html`?

Reply:
> I feel uneasy about the `\/` escaping as the only barrier here. Is there any escaping function WordPress ships that would also escape the `<` and `>` characters? Maybe `esc_html`?

This is something I've thought about for a long time. If you're interpolating directly into JS, I don't think there is a stronger escaping function than json_encode. If you can break out of that, we need to open a bug report on PHP.net because JSON by definition should be valid Javascript.

I worry that adding extra measures around it could unintentionally weaken this in a non obvious way.

How would you feel about casting the input to JSON_encode to `(string)`?

Reply:
> JSON by definition should be valid Javascript.

The caveat here is that we're in an HTML document and need to worry about more escape sequences than just JavaScript ones. That's what `esc_html` takes care of.

Reply:
Also, I'm curious what's some prior art on this in other WordPress plugins where this kind of data passing is happening.

Reply:
@adamziel It looks like the WP functions have a lot of moving parts where things could go wrong. How does this look?

https://github.com/WordPress/playground-tools/pull/124/files#diff-72b9ba31c538d25a3db28fcc6b6f58ca9b1fca1f66bab3947d565eaf85bbff07R79

Reply:
Sure, `htmlspecialchars` seems tempting here. At the same time, some WordPress functions exist specifically because of deficiencies of these escaping functions provided by PHP. What are some of the largest existing WordPress plugins doing in this particular situation and how can that inform the escaping strategy in this place and in all similar places across the board?

Reply:
@adamziel While I strongly agree that finding a solid, standard way of doing this is important, I think its exceedingly unlikely that someone would have XSS code in their administrative username. Would you agree that we can ship this as it stands, and iterate toward a more perfect solution in the future?

Reply:
Updated in 1ee4c4e97e29015cbc5f8e3f751012dd69c76344 to [use WordPress escaping.](https://developer.wordpress.org/apis/security/escaping/)